node/src/chain/fsm.rs

```diff
@@ -947,6 +1018,78 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network>
 
         Ok(false)
     }
+
+    async fn request_missing_block(&self, height: u64) {
+        let mut inv = Inv::new(0);
+        inv.add_block_from_height(height);
+        let get_resource =
+            GetResource::new(inv, Some(self.this_peer), u64::MAX, 1);
+
+        debug!(event = "request block", height, mode = "out_of_sync",);
+        if let Err(e) = self
+            .network
+            .read()
+            .await
+            .send_to_alive_peers(get_resource.into(), 2)
```

Comment:
If I didn't received from that peer, means that message got lost
If I request it again to the same peer if will be sent when the current batch is over, better to ask to someone else

Reply:
I agree with this argument. However, the same reasoning should also apply to the timeout event.
Most of the times, the timeout will expire due to a lost block (since we reset the timeout only upon acceptance).

In other words, I think we should request lost blocks to other peers, and use `get_missing_blocks` when approaching intermediate target (last_request), that is when we deuqued all requested blocks.